### DISCUSSION TITLE
403:mg_error        385 0x6 tls hs:

### DISCUSSION THREAD

**Discussion by schilton** [cesanta/mongoose#2222] [Ask questions, get answers] [answered]
I am running the HTTP-Restful-Server example with a slight modification that Sergio Caprile kindly provided.  The server runs well for awhile and then it crashes with the following error:

Last successful message:
99acf9 3 mongoose.c:5739:mg_tls_init    385 SSL accept OK

Error message:
99ae20 1 mongoose.c:403:mg_error        385 0x6 tls hs: rc -1, errSegmentation fault (core dumped)

Would someone please explain to me what this message is telling me and how I should go about resolving it.  Thank you very much for any insights you may provide.

**Comment by scaprile** [comment]
> I am running the HTTP-Restful-Server example with a slight modification that Sergio Caprile kindly provided. The server runs well for awhile and then it crashes with the following error:
> 
> Last successful message: 99acf9 3 mongoose.c:5739:mg_tls_init 385 SSL accept OK
> 
> Error message: 99ae20 1 mongoose.c:403:mg_error 385 0x6 tls hs: rc -1, errSegmentation fault (core dumped)
> 
> Would someone please explain to me what this message is telling me and how I should go about resolving it. Thank you very much for any insights you may provide.

`Segmentation fault (core dumped)` tells me that there is a high probability that you are on Linux
`mongoose.c:403:mg_error 385 0x6 tls hs: rc -1` : mg_error was called by some function in order to close the connection. The reason code is `-1` and that was during TLS handshake, you can search for that `tls hs: rc` and you'll find the exact place in the code.

The rest, I don't have the slightest idea of what you are doing. Your TLS connection is not establishing, that's all I can guess.

**Reply by schilton** [reply]
Sergio, you are correct that this is running on a linux distro, specifically Ubuntu 22.04.2 LTS.  I did a search as you suggested and the error occurs in void mg_tls_handshake(struct mg_connection *c).  I've attached to code, but the only thing we did differently from the example is to force the get request over https instead of http.  I know that a segmentation fault occurs when trying to improperly access memory, but where in the handshake process should I look?
[main_c.txt](https://github.com/cesanta/mongoose/files/11601899/main_c.txt)

**Reply by scaprile** [reply, marked as answer]
Works for me in Centos Stream 8 and Ubuntu 20.04 with OpenSSL (I also guess you are using openSSL)
BTW, I'm using our own certificates present on the repo, try with them and let us know
If it still fails to work, please indicate the TLS library and version you are using and we'll try to reproduce it.

**Reply by schilton** [reply]
Sergio,  I recompiled the app with MBEDTLS and am still receiving segmentation faults.  They do not appear to occur at the same point in the code and the server crashes are intermittent.  Attached is the cellp error log, the server code and the nmap results for the tls versions.
[tls_versions.txt](https://github.com/cesanta/mongoose/files/11643286/tls_versions.txt)
[main.txt](https://github.com/cesanta/mongoose/files/11643290/main.txt)
[cellp.log](https://github.com/cesanta/mongoose/files/11643291/cellp.log)

**Reply by scaprile** [reply]
That is too much work for me and sorry, I am a very lazy person.
Please:
> BTW, I'm using our own certificates present on the repo, try with them and let us know
If it still fails to work, please indicate the TLS library and version you are using and we'll try to reproduce it.

E.g: _I'm using mbedTLS x.x and/or openSSL x/x from the Ubuntu 22.04.02 LTS repo or downloaded myself
I tried your certificates and got the same problem (so there is no problem with mine)_

We are currently downloading a specific version of MbedTLS and building it ourselves, I suggest you see https://github.com/cesanta/mongoose/tree/master/examples/device-dashboard and try yourself doing the same

Our examples work in our testbeds, if you don't tell us exactly what you are doing, we can't guess
Please don't attach files disguised as text, zip your whole project and attach it, or push it to a repo and share the repo.
+ I strongly suggest you open an issue and fill the issue template when you do it.

**Comment by cpq** [comment]
@schilton @scaprile this question does not belong here - the right place for it is in the issues. There is nothing to discuss.

I'd like to ask to file an issue and delete such questions from the discussions - though not sure if the question can be deleted. Maybe we can add a "bin" category and move questions to the bin.
